benchmarks/advection_reaction_3D/raja/rhs3D.hpp

```diff
@@ -0,0 +1,606 @@
+
+  /* iterate over domain interior, computing advection */
+  if (c > 0.0)
+  {
+    /* flow moving in the positive x,y,z direction */
+    auto range = RAJA::make_tuple(RAJA::RangeSegment(1, nxl),
+                                  RAJA::RangeSegment(1, nyl),
+                                  RAJA::RangeSegment(1, nzl));
+
+    RAJA::kernel<XYZ_KERNEL_POL>(range,
+      [=] DEVICE_FUNC (int i, int j, int k) {
+      const realtype u_ijk = Yview(i,j,k,0);
+      const realtype v_ijk = Yview(i,j,k,1);
+      const realtype w_ijk = Yview(i,j,k,2);
+
+      // grad * u
+      dYview(i,j,k,0) =  cz * (u_ijk - Yview(i,j,k-1,0)); // du/dz
+      dYview(i,j,k,0) += cy * (u_ijk - Yview(i,j-1,k,0)); // du/dy
+      dYview(i,j,k,0) += cx * (u_ijk - Yview(i-1,j,k,0)); // du/dx
+
+      // grad * v
+      dYview(i,j,k,1) =  cz * (v_ijk - Yview(i,j,k-1,1)); // dv/dz
+      dYview(i,j,k,1) += cy * (v_ijk - Yview(i,j-1,k,1)); // dv/dy
+      dYview(i,j,k,1) += cx * (v_ijk - Yview(i-1,j,k,1)); // dv/dx
+
+      // grad * w
+      dYview(i,j,k,2) =  cz * (w_ijk - Yview(i,j,k-1,2)); // dw/dz
+      dYview(i,j,k,2) += cy * (w_ijk - Yview(i,j-1,k,2)); // dw/dy
+      dYview(i,j,k,2) += cx * (w_ijk - Yview(i-1,j,k,2)); // dw/dx
+    });
+  }
+  else if (c < 0.0)
+  {
+    /* flow moving in the negative x,y,z direction */
+    auto range = RAJA::make_tuple(RAJA::RangeSegment(0, nxl-1),
+                                  RAJA::RangeSegment(0, nyl-1),
+                                  RAJA::RangeSegment(0, nzl-1));
+    RAJA::kernel<XYZ_KERNEL_POL>(range,
+      [=] DEVICE_FUNC (int i, int j, int k) {
+      const realtype u_ijk = Yview(i,j,k,0);
+      const realtype v_ijk = Yview(i,j,k,1);
+      const realtype w_ijk = Yview(i,j,k,2);
+
+      // grad * u
+      dYview(i,j,k,0) =  cz * (Yview(i,j,k+1,0) - u_ijk); // du/dz
+      dYview(i,j,k,0) += cy * (Yview(i,j+1,k,0) - u_ijk); // du/dy
+      dYview(i,j,k,0) += cx * (Yview(i+1,j,k,0) - u_ijk); // du/dx
+
+      // grad * v
+      dYview(i,j,k,1) =  cz * (Yview(i,j,k+1,1) - v_ijk); // dv/dz
+      dYview(i,j,k,1) += cy * (Yview(i,j+1,k,1) - v_ijk); // dv/dy
+      dYview(i,j,k,1) += cx * (Yview(i+1,j,k,1) - v_ijk); // dv/dx
+
+      // grad * w
+      dYview(i,j,k,2) =  cz * (Yview(i,j,k+1,2) - w_ijk); // dw/dz
+      dYview(i,j,k,2) += cy * (Yview(i,j+1,k,2) - w_ijk); // dw/dy
+      dYview(i,j,k,2) += cx * (Yview(i+1,j,k,2) - w_ijk); // dw/dx
+    });
+  }
+
+  /* finish exchanging boundary information */
+  SUNDIALS_MARK_BEGIN(udata->prof, "Neighbor Exchange");
+  retval = udata->grid->ExchangeEnd();
+  if (check_retval(&retval, "ExchangeEnd", 1, udata->myid))
+    return(-1);
+  SUNDIALS_MARK_END(udata->prof, "Neighbor Exchange");
+
+
+  /* compute advection at process boundaries */
+  if (c > 0.0)
+  {
+    /* Flow moving in the positive x,y,z direction:
+    *  boundaries are west face, south face, back face */
+
+    /*   Create 3D views of receive buffers */
```

Comment:
Do we need to check `NDIMS` here also?

Reply:
Unfortunately, this is a more difficult fix.  If the flow is less than 3D, some parts of each kernel must be omitted, either resulting in (a) two `if` statements within each iteration of the forall (likely killing performance), or (b) separate 1D, 2D and 3D copies of each kernel (making things uglier and less maintainable).

I'll go with the latter of the two.